docs/rholang/rholangtut.md

```diff
@@ -0,0 +1,501 @@
+
+## Dining philosophers and deadlock
+
+     1 new philosopher1, philosopher2, north, south, knife, spoon in {
+     2   north!(*knife) |
+     3   south!(*spoon) |
+     4   for (@knf <- north) { for (@spn <- south) {
+     5     philosopher1!("Complete!") |
+     6     north!(knf) |
+     7     south!(spn)
+     8   } } |
+     9   for (@spn <- south) { for (@knf <- north) {
+    10     philosopher2!("Complete!") |
+    11     north!(knf) |
+    12     south!(spn)
+    13   } }
+    14 }
+
+The dining philosophers problem has two philosophers that share only one set of silverware.  Philosopher1 sits on the east side of the table while Philosopher2 sits on the west. Each needs both a knife and a spoon in order to eat.  Each one refuses to relinquish a utensil until he has used both to take a bite.  If both philosophers reach first for the utensil at their right, both will starve: Philosopher1 gets the knife, Philosopher2 gets the spoon, and neither ever lets go.
+
+Here's how to solve the problem:
+
+     1 new philosopher1, philosopher2, north, south, knife, spoon in {
+     2   north!(*knife) |
+     3   south!(*spoon) |
+     4   for (@knf <- north; @spn <- south) {
+     5     philosopher1!("Complete!") |
+     6     north!(knf) |
+     7     south!(spn)
+     8   } |
+     9   for (@spn <- south; @knf <- north) {
+    10     philosopher2!("Complete!") |
+    11     north!(knf) |
+    12     south!(spn)
+    13   }
+    14 }
+
+4, 9) The join operator, denoted with a semicolon `;`, declares that the continuation should only proceed if there is a message available on each of the channels simultaneously, preventing the deadlock above.
+
+## Crypto channels
+
+### Hashing
+
+There are three hashing functions available:
+- keccak256
+- sha256
+- blake2b256
+
+Hashing functions are exposed as channels which expect two arguments:
+- byte array to hash
+- return channel for sending back the hash
+
+
+#### Example usage:
+
+```rholang
+new x,y in {
+    x!(@"name"!("Joe") | @"age"!(40)) |  // (1)
+        for (@r <- x) {
+            @"keccak256Hash"!(r.toByteArray(), *y) // hash the program from (1)
+        } |
+        for (@h <- y) {
+            // the h here is hash of the rholang term we sent to the hash channel
+            // we can do anything we want with it but we choose to just print it
+            @"stdout"!(h)  // print out the keccak256 hash
+        }
+}
+```
+
+
+### Verify
+
+1. Let's hash some rholang program and print out it in base16. In rholang:
+```rholang
+new x,y in { 
+   x!(@"name"!("Joe") | @"age"!(40)) |  // (1)
+   for (@r <- x) { @"keccak256Hash"!(r.toByteArray(), *y) } |  // hash the program from (1)
+   for (@h <- y) { @"stdout"!(h) }  // print out the keccak256 hash
+}
+```
+
+This will print the hash of our program `(1)` : 
+`a6da46a1dc7ed715d4cd6472a736249a4d11142d160dbef9f20ae493de908c4e`
+
+2. We need pair of keys, let's generate some with `Ed25519` available in the project. In scala console:
```

Comment:
We need a pair of keys; let's generate some with `Ed25519`, available in the project. In the scala console:

Reply:
Done.